proof/access-control/Retype_AC.thy

```diff
@@ -337,125 +156,138 @@ lemma sita_detype:
   apply (auto simp: detype_def  intro: state_irqs_to_policy_aux.intros split: if_split_asm)
   done
 
-lemma sata_detype:
-  "state_asids_to_policy aag (detype R s) \<subseteq> state_asids_to_policy aag s"
-  apply (clarsimp)
-  apply (erule state_asids_to_policy_aux.induct)
-  apply (auto intro: state_asids_to_policy_aux.intros split: if_split_asm)
-  done
-
 (* FIXME: move *)
 lemmas pas_refined_by_subsets = pas_refined_state_objs_to_policy_subset
 
+lemma dtsa_detype: "domains_of_state (detype R s) \<subseteq> domains_of_state s"
+  by (auto simp: detype_def detype_ext_def
+          intro: domtcbs
+          elim!: domains_of_state_aux.cases
+          split: if_splits)
+
+
+locale Retype_AC_1 =
+  fixes aag :: "'a PAS"
+  assumes state_vrefs_detype[simp]:
+    "state_vrefs (detype R s) = (\<lambda>x. if x \<in> R then {} else state_vrefs s x)"
+  and sata_detype:
+    "state_asids_to_policy aag (detype R s) \<subseteq> state_asids_to_policy aag s"
+  and untyped_min_bits_ge_2:
+    "2 \<le> untyped_min_bits"
+  and clas_default_cap:
+    "\<And>tp. tp \<noteq> ArchObject ASIDPoolObj \<Longrightarrow> cap_links_asid_slot aag l (default_cap tp p sz dev)"
+  and cli_default_cap:
+    "\<And>tp. tp \<noteq> ArchObject ASIDPoolObj \<Longrightarrow> cap_links_irq aag l (default_cap tp p sz dev)"
+  and aobj_refs'_default':
+    "\<And>tp. is_aligned p (obj_bits_api (ArchObject tp) n)
+           \<Longrightarrow> aobj_ref' (arch_default_cap tp p n dev) \<subseteq> ptr_range p (obj_bits_api (ArchObject tp) n)"
+  and init_arch_objects_pas_refined:
+    "\<And>tp. \<lbrace>pas_refined aag and post_retype_invs tp refs
+                            and (\<lambda>s. \<forall>x\<in>set refs. x \<notin> global_refs s)
+                            and K (\<forall>ref \<in> set refs. is_aligned ref (obj_bits_api tp obj_sz))\<rbrace>
+           init_arch_objects tp ptr bits obj_sz refs
+           \<lbrace>\<lambda>_. pas_refined aag\<rbrace>"
+  and dmo_freeMemory_invs:
+    "do_machine_op (freeMemory ptr bits) \<lbrace>\<lambda>s :: det_ext state. invs s\<rbrace>"
+  and init_arch_objects_pas_cur_domain[wp]:
+    "\<And>tp. init_arch_objects tp ptr n us refs \<lbrace>pas_cur_domain aag\<rbrace>"
+  and retype_region_pas_cur_domain[wp]:
+    "\<And>tp. retype_region ptr n us tp dev \<lbrace>pas_cur_domain aag\<rbrace>"
+  and reset_untyped_cap_pas_cur_domain[wp]:
+    "reset_untyped_cap src_slot \<lbrace>pas_cur_domain aag\<rbrace>"
```

Comment:
Is there are reason these need to be proven in the `Arch` context?

Reply:
The lemmas which autosolve these are arch-specific (proved in a crunch at the start of ArchDetSchedAux_AI) and aren't requalified